### PR TITLE
Return read-only collections from `@cached` methods

### PR DESCRIPTION
**Pull request by squahtx**
It's important that collections returned from `@cached` methods are not
modified, otherwise future retrievals from the cache will return the
modified collection.

This applies to the return values from `@cached` methods and the values
inside the dictionaries returned by `@cachedList` methods. It's not
necessary for the dictionaries returned by `@cachedList` methods
themselves to be read-only.

Signed-off-by: Sean Quah <seanq@matrix.org>

### CODE REVIEW

File: synapse/handlers/room.py
Comment (squahtx) [inline synapse/handlers/room.py:1877]: Having to take a copy here is really unfortunate. We need something JSON-serializable here and the new `Sequence[...]` type hint doesn't guarantee that.

File: synapse/handlers/sync.py
Comment (squahtx) [inline synapse/handlers/sync.py:1382]: This goes straight into the `SyncResult` and is pretty much only going to be JSON-serialized from here on.
Having to take a copy to make mypy happy is unfortunate.

File: synapse/handlers/sync.py
Comment (squahtx) [inline synapse/handlers/sync.py:1624]: This looks like a bug. `global_account_data` comes from a cache.

File: synapse/storage/databases/main/devices.py
Comment (squahtx) [inline synapse/storage/databases/main/devices.py:712]: `get_cached_devices_for_user` returns a cached result.

We can't guarantee that the branch above won't be taken later on, so we have to take a copy here. Which is unfortunate.

File: synapse/handlers/directory.py
Comment (clokep) [inline synapse/handlers/directory.py:529]: I recommend using `Collection[str]` here if possible since `str` is a `Sequence[str]`, so it could let bugs slip in.

Comment (clokep) [reply synapse/handlers/directory.py:529]: (Looks like this might apply in a few places below.)

Comment (DMRobertson) [reply synapse/handlers/directory.py:529]: `str` is also a subtype of `Collection[str]`, I think?

Comment (DMRobertson) [reply synapse/handlers/directory.py:529]:
```
$ cat temp.py 
from typing import Collection
x: Collection[str]
x = "123"

$ mypy --strict temp.py 
Success: no issues found in 1 source file
```

Comment (clokep) [reply synapse/handlers/directory.py:529]: Right -- we need a `FrozenList` sort of class.

Comment (DMRobertson) [reply synapse/handlers/directory.py:529]: For cross-referencing, https://github.com/python/typing/issues/256. https://github.com/python/typing/issues/510 and https://github.com/python/typing/issues/256#issuecomment-917601678 are interesting ideas though.

Comment (squahtx) [reply synapse/handlers/directory.py:529]: https://stackoverflow.com/questions/11142397/does-python-have-an-immutable-list suggests that tuples can be considered to be `FrozenList`s. We could copy all these lists into tuples? (which I'm not a fan of)

Alternatively it may be possible to define some sort of `ReadOnlyList[T]` protocol which is somehow only satisfied by a real list (`__class__` has to be `Type[List[T]]` maybe?) but doesn't expose any methods to mutate it. However, that doesn't feel very python-y.

Any thoughts on which way we should go?

Comment (DMRobertson) [reply synapse/handlers/directory.py:529]: Just my opinions in passing:

- I don't mind tuples as frozen sequence types! But I imagine we'd end up spending a non-negligible chunk of time converting lists to tuples
- Bluntly: I wouldn't block this on trying to fix/improve Python's type system.
- _Maybe_ we could do something clever with a mypy plugin, but I think that's [a rabbit hole](https://youtu.be/rVJHQ7SkgwI?t=3).

**Review by DMRobertson — COMMENTED** [review]
I haven't looked through the full changes (sorry), but I'd love to see us land some of this: particularly bugfixes or things that don't incur performance costs. (Though it's Python: if you're worried about performance costs, it could be argued you're doing something wrong.)

One thing I'm not clear on is enforcement. Is there some way we can annotate @cached or some mypy plugin witchcraft we can do to which will force us to only cache functions with immutable return values? (Possibly we could use `Hashable` as an approximation of an `Immutable` type?)

File: synapse/handlers/directory.py
Comment (DMRobertson) [reply synapse/handlers/directory.py:529]: > _Maybe_ we could do something clever with a mypy plugin, but I think that's [a rabbit hole](https://youtu.be/rVJHQ7SkgwI?t=3).

I treated myself. See https://github.com/matrix-org/synapse/pull/14911

**Comment by DMRobertson** [comment]
> Is there some way we can annotate @cached or some mypy plugin witchcraft we can do to which will force us to only cache functions with immutable return values? (Possibly we could use `Hashable` as an approximation of an `Immutable` type?)

I indulged myself. https://github.com/matrix-org/synapse/pull/14911 is a mypy plugin which hopefully detects a bunch of these.

**Comment by DMRobertson** [comment]
I've tried to merge in develop to resolve the merge conflicts. (I'd like to see this land).

(And I also want to see what #14911 makes of it)

**Comment by DMRobertson** [comment]
> I've tried to merge in develop to resolve the merge conflicts. (I'd like to see this land).
> 
> (And I also want to see what #14911 makes of it)

Against develop: 181 errors
With this PR: 72 errors (some of which may be my hacky plugin).

File: synapse/handlers/directory.py
Comment (DMRobertson) [reply synapse/handlers/directory.py:532]: Since this comment we now have `StrCollection` as our home-grown workaround for this problem.

I think we should punt this problem to https://github.com/matrix-org/synapse/issues/14809 and try to land this PR as-is.

File: synapse/handlers/receipts.py
Comment (DMRobertson) [inline synapse/handlers/receipts.py:192]: General comment: I wonder if we ought to use `JsonMapping` more (The sequence may be immutable, but I can still use it to get references to mutable data and e.g. pop a key-value pair.)

For cached methods specifically, this is the kind of thing that https://github.com/matrix-org/synapse/pull/14911 can help with in the future

File: synapse/handlers/room.py
Comment (DMRobertson) [reply synapse/handlers/room.py:1932]: Agreed, though the cost of the copy should be low here.

File: synapse/handlers/sync.py
Comment (DMRobertson) [reply synapse/handlers/sync.py:1498]: At least it's a shallow copy :shrug:

File: synapse/push/push_rule_evaluator.py
Comment (DMRobertson) [inline synapse/push/push_rule_evaluator.py:1]: Is this file supposed to be here? I may have screwed up when trying to merge in develop

File: synapse/storage/databases/main/end_to_end_keys.py
Comment (DMRobertson) [inline synapse/storage/databases/main/end_to_end_keys.py:945]: Why do we need the cast here---because Dict is invariant?

File: synapse/storage/databases/main/receipts.py
Comment (DMRobertson) [inline synapse/storage/databases/main/receipts.py:293]:
```suggestion
    ) -> Sequence[JsonDict]:
```

for consistency with `_get_linearized_receipts_for_room` below

File: synapse/storage/databases/main/receipts.py
Comment (DMRobertson) [inline synapse/storage/databases/main/receipts.py:359]: I think this `Dict` is okay, because it comes from `@cachedList`?

File: synapse/storage/databases/main/roommember.py
Comment (DMRobertson) [inline synapse/storage/databases/main/roommember.py:164]: Guessing this came in from a merge?

File: synapse/storage/databases/main/signatures.py
Comment (DMRobertson) [inline synapse/storage/databases/main/signatures.py:29]: Was this wrong to have the extra level of `Dict[str, ...]` nesting before? If so, I wonder why we didn't spot it until now.

File: synapse/storage/databases/main/receipts.py
Comment (clokep) [reply synapse/storage/databases/main/receipts.py:359]: Yes, `cachedList` doesn't do the caching. 👍 The values in the dict need to be non-mutable.

**Review by DMRobertson — COMMENTED** [review]
Huge thanks for this Sean. I think this is a great change and am looking forward to being shot at in the foot less. Just a few things to check:

File: synapse/push/push_rule_evaluator.py
Comment (clokep) [reply synapse/push/push_rule_evaluator.py:1]: That file was deleted.

File: synapse/handlers/directory.py
Comment (clokep) [reply synapse/handlers/directory.py:489]: I think we should break out any bugfixes we find to separate PRs -- a big refactoring like this including bugfixes is a bit scary?

File: synapse/handlers/sync.py
Comment (clokep) [reply synapse/handlers/sync.py:1526]: I'd be curious of the error here? Is something in *our* code unhappy or is it a type hint of like `json.dump`?

I think it'd be OK to add a `cast` with a comment explaining it is for performance.

File: synapse/handlers/sync.py
Comment (clokep) [reply synapse/handlers/sync.py:1770]: As above, I think we should just fix this separately? (I also noticed this as part of #14973.)

File: synapse/handlers/receipts.py
Comment (clokep) [reply synapse/handlers/receipts.py:192]: I think so

I suggest that we:

1. Don't do that here.
2. Rebase #14911 after merging this.
3. Fix-up interior values to be non-mutable (including using `JsonMapping`).

File: synapse/storage/databases/main/end_to_end_keys.py
Comment (clokep) [reply synapse/storage/databases/main/end_to_end_keys.py:945]: @squahtx Any memory of this?

File: synapse/storage/databases/main/roommember.py
Comment (clokep) [reply synapse/storage/databases/main/roommember.py:165]: Meaning it should be deleted?

File: synapse/storage/databases/main/signatures.py
Comment (clokep) [reply synapse/storage/databases/main/signatures.py:29]: I suspect we should fix this separately if it was completely wrong?

File: synapse/handlers/directory.py
Comment (clokep) [reply synapse/handlers/directory.py:489]: #15038

File: synapse/storage/databases/main/devices.py
Comment (clokep) [reply synapse/storage/databases/main/devices.py:786]: This code seems kind of confusing in the first place -- it doesn't make sense to have a user with device IDs & all devices. I think some refactoring will help.

Comment (clokep) [reply synapse/storage/databases/main/devices.py:786]: #15040.

File: synapse/storage/databases/main/end_to_end_keys.py
Comment (squahtx) [reply synapse/storage/databases/main/end_to_end_keys.py:945]: I can't remember exactly, but it's likely due to dict invariance.
`_get_e2e_cross_signing_signatures_txn` returns a `Dict[str, Optional[Dict[str, JsonDict]]]`, but this PR wants it to return a `Dict[str, Optional[***Mapping***[str, JsonDict]]]`

File: synapse/storage/databases/main/signatures.py
Comment (squahtx) [reply synapse/storage/databases/main/signatures.py:29]: The reasoning was that if the `@cached` method returns a `T`, the `@cachedList` method (below) must return a `Dict / Mapping[Key, T]`, and we infer `T` to be a `Dict / Mapping[str, bytes]`.

We never call the non-list method so this error no impact.

File: synapse/handlers/sync.py
Comment (squahtx) [reply synapse/handlers/sync.py:1526]: It's a type hint problem. Only `List`s and `Tuple`s are JSON-serializable, except they're mutable types. I would have updated the definition of `SyncResult` otherwise.

File: synapse/storage/databases/main/roommember.py
Comment (squahtx) [reply synapse/storage/databases/main/roommember.py]: I've gone and fixed the merge.

File: synapse/storage/databases/main/receipts.py
Comment (squahtx) [reply synapse/storage/databases/main/receipts.py]: done

File: synapse/handlers/sync.py
Comment (squahtx) [reply synapse/handlers/sync.py:1770]: PRed as https://github.com/matrix-org/synapse/pull/15047

**Review by DMRobertson — APPROVED** [review]
LGTM!

**Comment by squahtx** [comment]
I had to fix some merge conflicts.

**Review by DMRobertson — APPROVED** [review]

**Pull request merged**
